app/bundles/LeadBundle/Controller/LeadController.php

```diff
@@ -2120,4 +2124,112 @@ public function contactStatsAction(int $objectId)
+        if (null === $lead
+            || !$this->security->hasEntityAccess(
+                'lead:leads:editown',
+                'lead:leads:editother',
+                $lead->getPermissionUser()
+            )
+        ) {
+            return $this->accessDenied();
+        }
+
+        $pointGroups = $pointGroupModel->getEntities();
+        $fields      = [];
+        foreach ($pointGroups as $group) {
+            $fields[] = ContactGroupPointsType::getFieldKey($group->getId());
+        }
+
+        $initData = [];
+        foreach ($lead->getGroupScores() as $score) {
+            $initData[ContactGroupPointsType::getFieldKey($score->getGroup()->getId())] = $score->getScore();
+        }
+
+        $form = $this->formFactory->create(ContactGroupPointsType::class, $initData, [
+            'action' => $this->generateUrl('mautic_contact_action', [
+                'objectAction' => 'contactGroupPoints',
+                'objectId'     => $objectId,
+            ]),
+            'point_groups' => $pointGroups,
+        ]);
+
+        if (Request::METHOD_POST === $request->getMethod()) {
+            if (!$this->isFormCancelled($form)) {
+                if ($this->isFormValid($form)) {
+                    $leadUpdated = false;
+                    $postData    = $form->getData();
+                    foreach ($pointGroups as $group) {
+                        $scoreKey = ContactGroupPointsType::getFieldKey($group->getId());
+                        $oldScore = $initData[$scoreKey] ?? null;
+                        $newScore = $postData[$scoreKey];
+                        if (!is_null($oldScore) && is_null($newScore)) {
+                            // set 0 when the new score is not present, but the record exists
+                            $newScore = 0;
+                        }
+
+                        if (!is_null($newScore) && $newScore !== $oldScore) {
+                            $pointGroupModel->adjustPoints($lead, $group, $newScore, Lead::POINTS_SET);
+                            $delta = $newScore - ($oldScore ?? 0);
+
+                            // add a lead point change log
+                            $log = new PointsChangeLog();
+                            $log->setDelta($delta);
+                            $log->setLead($lead);
+                            $log->setType('manual');
+                            $log->setEventName($this->translator->trans('mautic.point.event.manual_change'));
+                            $log->setActionName('');
+                            $log->setIpAddress($ipLookupHelper->getIpAddress());
+                            $log->setDateAdded(new \DateTime());
+                            $log->setGroup($group);
+                            $lead->addPointsChangeLog($log);
+                            $leadUpdated = true;
+                        }
+                    }
+
+                    if ($leadUpdated) {
+                        $model->saveEntity($lead);
```

Comment:
Perhaps this line could be called instead of line 2194 and then the `$leadUpdated` variable can be deleted?

Reply:
I use this variable for better optimisation, the save method is only called once, no matter how many point groups we add or update. Do you see any reason, other than code readability, to move the entity save to a loop?

Reply:
Makes sense. Should have see that. Thanks!